synapse/storage/databases/main/events_worker.py

```diff
@@ -1928,6 +1928,18 @@ def is_event_next_to_gap_txn(txn: LoggingTransaction) -> bool:
                 LIMIT 1
             """
 
+            # We consider any forward extremity as the latest in the room and
+            # not a forward gap.
+            #
+            # To expand, even though there is technically a gap at the front of
+            # the room where the forward extremities are, we consider those the
+            # latest messages in the room so asking other homeservers for more
+            # is useless. The new latest messages will just be federated as
+            # usual.
+            txn.execute(forward_extremity_query, (event.room_id, event.event_id))
+            if txn.fetchone():
+                return False
```

Comment:
I'm moving this up (from lines 1948-1960) while I'm here so that it's next to the query definition. I found it it confusing that we defined two queries, then didn't use them for a while.

Also simplifying the condition to avoid building a result list.